runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,227 @@
+        public void run() {
+            try {
+                log.info("{}:Seeking txStream to {}", listener.toString(), startAddress + 1);
+                txnStream.seek(startAddress + 1);
+            } catch (Throwable throwable) {
+                log.warn("{}:onError at seek to {}", listener.toString(), startAddress + 1);
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            try {
+                while (!shutdown.get()) {
+                    while (!txnStream.hasNext()) {
+                        log.trace("{}::run() all caught up. Taking a 50ms nap", listener.toString());
+                        int DEFAULT_TIMEOUT_IN_MILLI = 50;
+                        TimeUnit.MILLISECONDS.sleep(DEFAULT_TIMEOUT_IN_MILLI);
+                    }
+                    ILogData logData = txnStream.nextUpTo(Address.MAX);
+                    MultiObjectSMREntry multiObjSMREntry = (MultiObjectSMREntry) logData.getPayload(runtime);
+                    long epoch = logData.getEpoch();
+                    Map<TableSchema, List<CorfuStreamEntry>> entries = new HashMap<>();
+                    // first only filter by the stream IDs of interest that are present in this logData
+                    logData.getStreams().stream().filter(tablesOfInterest::containsKey)
+                            .forEach(streamId -> entries.put(tablesOfInterest.get(streamId),
+                                    // Only extract the list of updates per stream as a list
+                                    multiObjSMREntry.getSMRUpdates(streamId).stream().map(smrEntry ->
+                                            CorfuStreamEntry.fromSMREntry(smrEntry,
+                                                    epoch,
+                                                    tablesOfInterest.get(streamId).getKeyClass(),
+                                                    tablesOfInterest.get(streamId).getPayloadClass(),
+                                                    tablesOfInterest.get(streamId).getMetadataClass())
+                                    ).collect(Collectors.toList())));
+
+                    if (!entries.isEmpty()) {
+                        CorfuStreamEntries callbackResult = new CorfuStreamEntries(entries);
+                        log.debug("{}::onNext with {} updates", listener.toString(), entries.size());
+                        long onNextStart = System.nanoTime();
+                        listener.onNext(callbackResult);
+                        long onNextEnd = System.nanoTime();
+                        if (TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart) > 0) {
+                            log.debug("{}::onNext took {}s", listener.toString(),
+                                    TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart));
+                        }
+                    }
+                }
+            } catch (Throwable throwable) {
```

Comment:
looks like there is a problem here, you do not handle InterruptedException it means there could be problems when other threads will try to stop this thread. Intellij IDEA should makes this line yellow with the warning/explanation.

P.S. I know you execute interrupt() in `unsubscribe` method. I can't require you or claim but the way how this while cycle is written seems fragile to me

Reply:
StreamingIT.java has a test case that covers unsubscribe() and validate that it does not receive any future updates. That test seems to be working as expected.
But if you have an alternate better way please suggest. For now I am just relying on the fact that InterruptedException is also a type of Throwable
https://docs.oracle.com/javase/7/docs/api/java/lang/InterruptedException.html